polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -93,21 +93,45 @@ pub(super) fn temporal_range_dispatch(
     );
     const TO_MS: i64 = SECONDS_IN_DAY * 1000;
 
-    let rng_start = start.to_physical_repr();
-    let rng_stop = stop.to_physical_repr();
-    let dtype = start.dtype();
+    let start_dtype = start.dtype();
 
-    let mut start = rng_start.cast(&DataType::Int64)?;
-    let mut stop = rng_stop.cast(&DataType::Int64)?;
+    // Note: `start` and `stop` have already been cast to their supertype,
+    // so only `start`'s dtype needs to be matched against.
+    let (mut start, mut stop) = match start_dtype {
+        #[cfg(feature = "timezones")]
+        DataType::Datetime(_, Some(_)) => (
+            start
+                .datetime()
+                .unwrap()
+                .replace_time_zone(None, None)?
+                .into_series()
+                .to_physical_repr()
+                .cast(&DataType::Int64)?,
+            stop.datetime()
+                .unwrap()
+                .replace_time_zone(None, None)?
+                .into_series()
+                .to_physical_repr()
+                .cast(&DataType::Int64)?,
+        ),
+        _ => (
+            start.to_physical_repr().cast(&DataType::Int64)?,
+            stop.to_physical_repr().cast(&DataType::Int64)?,
+        ),
+    };
 
-    let (tu, tz) = match dtype {
```

Comment:
the output dtype isn't necessarily `start.dtype()` (if `tz` is not None). so instead of just recording `tu` and `tz`, I'm assigning to `dtype`, which contains `tu` and `tz`.

Then,
- for `Date`, I just use `Milliseconds` as `tu` and `None` for `tz`
- for `Time`, same, but with `Nanoseconds` as `tu`
- for `Datetime`, use its `tu` and `tz`